File: Java-Kotlin/src/main/java/jianzhi/jz26.java
package jianzhi;

import common.datastructure.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class jz26 {

    public TreeNode Convert(TreeNode root) {
        if (root == null) {
            return root;
        }
        if (root.left == null && root.right == null) {
            return root;
        }

        List<TreeNode> list = new ArrayList<>();
        inOrder(root, list);

        TreeNode pre = list.get(0);
        TreeNode dummy = pre;
        for (int i = 1; i < list.size(); i++) {
            TreeNode cur = list.get(i);
            pre.right = cur;
            cur.left = pre;
            pre = cur;
        }
        return dummy;
    }

    private static void inOrder(TreeNode node, List<TreeNode> list) {
        if (node == null) {
            return;
        }

        inOrder(node.left, list);
        list.add(node);
        inOrder(node.right, list);
    }
}

File: Go-Solutions/nowcoder/jz26.go
package nowcoder

func Convert(root *TreeNode) *TreeNode {
	// write code here

	if root == nil {
		return root
	}

	if root.Left == nil && root.Right == nil {
		return root
	}

	res := make([]*TreeNode, 0)
	inOrder26(root, &res)

	pre := res[0]
	dummy := pre
	for i := 1; i < len(res); i++ {
		cur := res[i]
		pre.Right = cur
		cur.Left = pre
		pre = cur
	}
	return dummy
}

func inOrder26(root *TreeNode, res *[]*TreeNode) {
	if root == nil {
		return
	}

	inOrder26(root.Left, res)
	*res = append(*res, root)
	inOrder26(root.Right, res)
}
